### PR TITLE
Improve performance of differentiable matrix calculations for Pauli Sentences

### PR DESCRIPTION
**Pull request by albi3ro**
**Context:**

The new operator arithmetic now enabled provides a fully differentiable pathway for calculating the matrix that was not available before.

Unfortunately, a differentiable calculation of a large dense matrix can be quite slow, and users may not realize they should be using a sparse matrix formulation instead. Just because we offer it does not make it a universally good idea.

This PR hopes to close some of that performance gap, by taking advantage of the structure of pauli sentences to improve the efficiency of calculating the matrix.

**Description of the Change:**

Adds in a new pathway for calculating the matrix of a pauli sentence.

**Benefits:**

Some performance numbers for the H5 molecule follow:

Screenshot 2024-04-25 at 9 57 05 AM

While the differentiable matrix is still not as fast as sparse matrix calculation, this method makes up a lot of ground. Note that the dense matrix is now on the same order of magnitude as the sparse matrix calculation for the legacy hamiltonian.

This formulation is also jit-able, and jitting the calculation of the matrix allows us to pick up another order of magnitude improvement.

**Possible Drawbacks:**

Dense is still not as fast as the sparse matrix, and might not scale as well as the sparse formulation.

Also, this pathway is much less intuitive and involves some numerical black magic.

**Related GitHub Issues:**

### CODE REVIEW

File: pennylane/pauli/pauli_arithmetic.py [hunk 1]
```diff
@@ -922,6 +885,25 @@ def _to_sparse_mat(self, wire_order, buffer_size=None):
         matrix.eliminate_zeros()
         return matrix
 
+    def _to_dense_mat(self, wire_order):
+        """Compute the sparse matrix of the Pauli sentence by efficiently adding the Pauli words
```
Comment (mudit2812) [inline pennylane/pauli/pauli_arithmetic.py:889]:
```suggestion
        """Compute the dense matrix of the Pauli sentence by efficiently adding the Pauli words
```

File: pennylane/pauli/pauli_arithmetic.py [hunk 2]
```diff
@@ -964,6 +946,24 @@ def _get_same_structure_csr(self, pauli_words, wire_order):
...
+        base_matrix.indptr = _cached_arange(
+            matrix_size + 1
+        )  # Non-zero entries by row (starting from 0)
+        base_matrix = base_matrix.toarray()
+        coeff = self[pauli_words[0]]
+        data = coeff * qml.math.convert_like(data0, coeff)
+        for pw in pauli_words[1:]:
+            data += self[pw] * pw._get_csr_data(wire_order, 1)
+
+        return qml.math.einsum("ij,i->ij", base_matrix, data)
```
Comment (vincentmr) [inline pennylane/pauli/pauli_arithmetic.py:965]: Smart, I like it.

**Review by vincentmr — APPROVED** [review]
Looks good to me.

**Comment by Qottmann** [comment]
[sc-58850]

File: pennylane/pauli/pauli_arithmetic.py (hunk 2 above)
Comment (albi3ro) [reply pennylane/pauli/pauli_arithmetic.py:965]: Well, large matrices and vectors is what we care about here.

File: tests/devices/qubit/test_apply_operation.py [hunk 3]
```diff
@@ -389,7 +389,8 @@ def test_large_state_small_matrix_evolves_matrix(self, mocker):
 
         # seems like _evolve_state_vector_under_parametrized_evolution calls
         # einsum twice, and the default apply_operation only once
-        assert spy.call_count == 1
+        # and it seems that getting the matrix from the hamiltonian calls einsum a few times.
+        assert spy.call_count == 6
```
Comment (trbromley) [inline tests/devices/qubit/test_apply_operation.py:393]: If we don't understand so well where these numbers come from, why are we testing them at all?

**Review by trbromley — COMMENTED** [review]
Thanks @albi3ro! A changelog entry would be very appreciated as this is a nice perf improvement.

File: tests/devices/qubit/test_apply_operation.py (hunk 3 above)
Comment (albi3ro) [reply tests/devices/qubit/test_apply_operation.py:393]: don't know why we were testing it, but now we call einsum more times due to it's use when calculating the matrix.

File: pennylane/pauli/pauli_arithmetic.py (hunk 2 above)
Comment (albi3ro) [reply pennylane/pauli/pauli_arithmetic.py:987]: Tried substituting `o2` and got a bunch of test failures.

Comment (albi3ro) [reply pennylane/pauli/pauli_arithmetic.py:987]: would `qml.math.einsum("i, ij ->ij", data, base_matrix)` be any faster?

Comment (vincentmr) [reply pennylane/pauli/pauli_arithmetic.py:987]: > would `qml.math.einsum("i, ij ->ij", data, base_matrix)` be any faster?

That appears like the correct `o2` one.

Comment (vincentmr) [reply pennylane/pauli/pauli_arithmetic.py:987]: I would think `einsum` is smart enough to handle both in an optimal way, such that they come out similar speed. `o1` should be a loop over scaling rows and `o2` should be a loop over vector-multiply. `o1` has fewer memory accesses in the limit of long rows. `o2` might be better for short rows.

Comment (albi3ro) [reply pennylane/pauli/pauli_arithmetic.py:987]: I'm just going to leave it be for now, but we could probably come back to this line at some point.

**Comment by trbromley** [comment]
Note that this supersedes https://github.com/PennyLaneAI/pennylane/pull/5392.

**Review by mudit2812 — COMMENTED** [review]
Looks good. I'm happy with the tests mostly, but looks like interface testing could be bumped up a bit as codecov is failing on the branches for casting for tensorflow

File: pennylane/pauli/pauli_arithmetic.py [hunk 4]
```diff
@@ -964,6 +956,36 @@ def _get_same_structure_csr(self, pauli_words, wire_order):
...
+        base_matrix.indices = pauli_words[0]._get_csr_indices(wire_order)
+        base_matrix.indptr = _cached_arange(
+            matrix_size + 1
+        )  # Non-zero entries by row (starting from 0)
+        base_matrix = base_matrix.toarray()
+        coeff = self[pauli_words[0]]
+        ml_interface = qml.math.get_interface(coeff)
+        if ml_interface == "torch":
+            data0 = qml.math.convert_like(data0, coeff)
+        elif ml_interface == "tf":
```
Comment (albi3ro) [inline pennylane/pauli/pauli_arithmetic.py:974]:
```suggestion
        elif ml_interface == "tensorflow":
```

File: pennylane/pauli/pauli_arithmetic.py [hunk 5]
```diff
@@ -964,6 +956,36 @@ def _get_same_structure_csr(self, pauli_words, wire_order):
...
+            data0 = qml.math.cast_like(data0, coeff)
+        data = coeff * data0
+        for pw in pauli_words[1:]:
+            coeff = self[pw]
+            csr_data = pw._get_csr_data(wire_order, 1)
+            ml_interface = qml.math.get_interface(coeff)
+            if ml_interface == "torch":
+                csr_data = qml.math.convert_like(csr_data, coeff)
+            elif ml_interface == "tf":
```
Comment (albi3ro) [inline pennylane/pauli/pauli_arithmetic.py:983]:
```suggestion
            elif ml_interface == "tensorflow":
```

**Review by mudit2812 — APPROVED** [review]

**Pull request merged**